halibot: merge the agents and modules dict into one

richteer opened the pull request:
For a long time, it has been debated if the agents and modules
dictionaries really needed to be separate. Both contain HalObjects with
similar interfaces, so they are not so different. However, with the
introduction of RIs, the leading argument against merging them has been
eliminated: agent and module instances can no longer share the same
name.

Since an agent and module can't have the same name, there is effectively
no longer a point to separating them. They have been replaced by a small
dictionary class that provides read-only attributes that handle the
filtering for you: .agents and .modules for geting just agents and
modules respectively.

This simplifies the lookup and addition of new objects, as there is no
longer two separate dictionaries that need to be examined.

richteer committed:
halibot: merge the agents and modules dict into one

For a long time, it has been debated if the agents and modules
dictionaries really needed to be separate. Both contain HalObjects with
similar interfaces, so they are not so different. However, with the
introduction of RIs, the leading argument against merging them has been
eliminated: agent and module instances can no longer share the same
name.

Since an agent and module can't have the same name, there is effectively
no longer a point to separating them. They have been replaced by a small
dictionary class that provides read-only attributes that handle the
filtering for you: .agents and .modules for geting just agents and
modules respectively.

This simplifies the lookup and addition of new objects, as there is no
longer two separate dictionaries that need to be examined.

sjrct commented on halibot/halibot.py:
```diff
@@ -130,10 +136,7 @@ def _instantiate_modules(self):
-		if name in self.modules: return self.modules[name]
-		if name in self.agents: return self.agents[name]
-		return None
+		return self.objects.get(name, None)
```
I don't remember who uses this (I don't think anyone outside of core though), but we probably should just completely remove `get_object`.

richteer replied:
Admittedly, I was too lazy to check. I will remove this and clear any uses of it in the next update

sjrct commented on halibot/halibot.py:
```diff
@@ -130,10 +136,7 @@ def _instantiate_modules(self):
 			self.add_module_instance(k, obj(self, conf))
```
Is there a reason for there to be a separate `add_module_instance` and `add_agent_instance` now? The only thing different is the log string.

Merging these would mean we can merge `_instantiate_agents` and `_instantiate_modules` too.

richteer replied:
No, since packages (and probably sooner) the loading logic is completely the same now. Will merge these functions in the next revision.

sjrct reviewed (changes requested):
Should work as is, but I would like to see `get_object` gone; `s/add_.*instance/add_instance/g`; and `s/_instantiate_.*\(self\)/_instantiate_objects(self, confkey)/g`. Or, ya know, something like that.

sjrct commented on halibot/halibot.py:
```diff
@@ -17,11 +17,18 @@
+			return dict(filter(lambda x: isinstance(x[1], HalModule), self.items()))
+		elif key == "agents":
+			return dict(filter(lambda x: isinstance(x[1], HalAgent), self.items()))
+		return super.__getattr__(self, key)
```
Maybe use of the `@property` decorator is cleaner?

richteer replied:
Definitely is. Thanks for pointing me to that decorator, that might come in handy in other places too.

richteer committed:
halibot: remove get_object method

Since agents and modules have been merged into one dictionary, this
patch removes the get_object() method, and replaces it with a direct
call to the objects' get() method.

richteer committed:
halibot: merge loading functions

Continuing on the trend of merging duplicate efforts, this commit
replaces add_*_instance with add_instance, since it is completely
independent of what type of object is being added.

The internal loading function for loading modules/agents from the config
has also been merged into _instantiate_objects, which now takes which
objects to load.

richteer commented:
Notable changes:
- `ObjectDict()` now uses `@property`
- merge of `add_module_instance()` and `add_agent_instance()`
- merge of `_instantiate_modules()` and `_instantiate_agents()`
- removed `get_object()`, replaced all calls with `.objects.get()`
